src/transformers/models/modernbert_decoder/modular_modernbert_decoder.py

```diff
@@ -0,0 +1,1305 @@
+            self.local_attention = (config.local_attention // 2, config.local_attention // 2)
+        else:
+            self.local_attention = (-1, -1)
+
+        rope_theta = config.global_rope_theta
+        if self.local_attention != (-1, -1):
+            if config.local_rope_theta is not None:
+                rope_theta = config.local_rope_theta
+
+        self.rotary_emb = ModernBertRotaryEmbedding(config=config, dim=self.head_dim, base=rope_theta)
+        self.Wo = nn.Linear(config.hidden_size, config.hidden_size, bias=config.attention_bias)
+        self.out_drop = nn.Dropout(config.attention_dropout) if config.attention_dropout > 0.0 else nn.Identity()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[tuple[torch.Tensor]] = None,
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        **kwargs,
+    ) -> torch.Tensor:
+        batch_size, seq_len, _ = hidden_states.shape
+
+        qkv = self.Wqkv(hidden_states)
+        qkv = qkv.view(batch_size, seq_len, 3, self.num_heads, self.head_dim)
+
+        # Fallback: create position_ids only if not provided
+        if position_ids is None:
+            device = hidden_states.device
+            if past_key_value is not None:
+                # For incremental decoding, start from cached length
+                cache_length = past_key_value[0].shape[-2]
+                position_ids = torch.arange(cache_length, cache_length + seq_len, dtype=torch.long, device=device)
+            else:
+                # For initial forward pass, start from 0
+                position_ids = torch.arange(seq_len, dtype=torch.long, device=device)
+            position_ids = position_ids.unsqueeze(0).expand(batch_size, -1)
+
+        # Ensure position_ids has the right shape if provided externally
+        if position_ids.dim() == 1:
+            position_ids = position_ids.unsqueeze(0).expand(batch_size, -1)
+
+        # Apply rotary embeddings
+        if past_key_value is not None:
+            # Extract query, key, value first for caching case
+            query, key, value = qkv.transpose(3, 1).unbind(dim=2)
+            # query, key, value: [batch_size, num_heads, seq_len, head_dim]
+
+            # create QKV format for RoPE computation
+            qkv_for_rope = torch.stack([query, key, value], dim=2).transpose(1, 2)
+            cos, sin = self.rotary_emb(qkv_for_rope, position_ids=position_ids)
+
+            # Apply RoPE only to the new query and key
+            query, key = apply_rotary_pos_emb(query, key, cos, sin)
+
+            # Concatenate with past keys and values
+            past_key, past_value = past_key_value
+            key = torch.cat([past_key, key], dim=-2)
+            value = torch.cat([past_value, value], dim=-2)
```

Comment:
let's use the `Cache` api that we have in transformers ! this is very restrictive